README.md
# PyODS
### An Online Disk Server written in python

The Python Online Disk Server allows you to serve cd/dvd/bluray optical drives and iso images to multiple Mac OS
clients on a network. PyODS shares (without credentials at present) all images found in a specified root folder,
and all optical drives attached to the system.

This emulates the network CD rom feature in Mac OS and allow other operating systems to provide access to disks
and disk images via this protocol. The client is found in the Remote Disc section of Finder. Multiple drives and
images can be shared at the same time, images can be uploaded and stored on the server via a RESTful API.

If you're looking for how to share your ubuntu cd drive to a mac, this is probably what you want.

For more information take a look at apple's support article here: https://support.apple.com/en-gb/HT203973

## Server API

### MacOS required API


    Method      HEAD
    Path        /<disk>.dmg   
    User-Agent  CCURLBS::statImage

Retrieves information regarding the disk size it works the same for images or optical drives attached
to the system.

    Method      GET
    Path        /<disk>.dmg
    Range       bytes=(\d*?)-(\d*)   
    User-Agent  CCURLBS::readDataFork

Retrieves a chunk of data from the disk, read as a block device.

NB: There are a number of other endpoints and file accesses which occur on the server, which the server
gracefully returns what is asked for or, a 404 error. This seems to work appropriately in most situations
however Apple may change their mind in the future.

### Extended Image API

    Method      GET
    Path        /images

List the disks on the server in JSON format

    Method      POST
    Path        /images
    Data        Multipart encoded file upload

Upload a file to the image storage, this will fail if the user does not have write access to that folder.

    Method      DELETE
    Path        /images/image.dmg

Delete an image file from the server.

### Extended JSON RPC API (in progress)

I've extended the MacOS API so we can do some nifty things in the future, like burning discs
ripping discs, uploading and sharing disc images on the local network, that sort of deal.
The goal here is to make CD/DVD media supportable in the future, and more useful than ever.

    Method      JSONRPC
    Path        /optical/<disk>
    RPC Method  open

Open the CD/DVD tray.

    Method      JSONRPC
    Path        /optical/<disk>
    RPC Method  close

Close the CD/DVD tray (if possible, slot drives won't always react).

    Method      JSONRPC
    Path        /burn/<disk image>/<disk target>
    RPC Method  burn

Tries to burn an image referenced by it's disk id to a target optical drive referenced by it's disk id

    Method      JSONRPC
    Path        /burn/<disk>
    RPC Method  erase

Tries to erase a CD-RW/DVD-RW found in the drive

    Method      JSONRPC
    Path        /backup/<disk>/<image>
    RPC Method  store

Backup an optical disc to an image file.

    Method      JSONRPC
    Path        /backup/<disk>/<callback url>/<target size>
    RPC Method  dvdencode

Backup a DVD movie to an MP4 file

    Method      JSONRPC
    Path        /backup/<disk>/<callback url>
    RPC Method  cdencode

Backup an audio cd to flac files


## To build a debian package

    dpkg-buildpackage -us -uc -b


debian/changelog
pyods (0.1-1) unstable; urgency=low

  * Initial public release

 -- Karl Lattimer <karl@qdh.org.uk>  Mon, 21 Oct 2019 09:22:00 +0000


pyods/tools.py
import os
import socket
from pathlib import Path
import re
from subprocess import Popen, STDOUT, PIPE


def get_local_ip():
    s = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
    s.connect(('10.255.255.255', 1))
    ip = s.getsockname()[0]
    s.close()
    return ip


def _is_image(filename):
    try:
        ext = os.path.splitext(filename)[1][1:].lower()
    except:
        return False

    if ext in ["iso", "img", "dmg"]:
        return True
    return False


def list_removable_drives():
    script = """grep -Hv ^0$ /sys/block/*/removable |
                sed s/removable:.*$/device\\/uevent/ |
                xargs grep -H ^DRIVER=sd |
                sed s/device.uevent.*$/size/ |
                xargs grep -Hv ^0$ |
                cut -d / -f 4"""


def list_images(path):
    files = list(Path(path).glob('**/*.*'))
    files = [f for f in files if _is_image(f)]
    files.sort()
    return files


def _getdev(line):
    try:
        dev = re.findall('dev=\'(.*?)\'', line)[0]
    except:
        return None
    return dev


def list_optical_drives():
    out = Popen(["wodim", "--devices"], stderr=STDOUT, stdout=PIPE)

    (output, returncode) = out.communicate()[0], out.returncode
    if returncode != 0:
        raise Exception("Failed to execute wodim command return code %d" % returncode)

    drives = [_getdev(l) for l in output.split()]
    drives = [d for d in drives if d is not None]
    drives.sort()
    return drives


pyods/__init__.py
from .OnlineDisk import DiskImage, OpticalDrive
from .tools import list_images, list_optical_drives, get_local_ip
import cherrypy
import logging
import json
import os
import importlib
import datetime


search_config = [
    "/etc/pyods/config.json",
    "~/.pyods/config.json",
    "/opt/venvs/pyods/config/config.json"
]


#
# Monkey patch for JSON Serialisation
#
def _default(self, obj):
    if type(obj) == datetime.datetime:
        return obj.isoformat()
    try:
        out = getattr(obj.__class__, "serialise", _default.default)(obj)
    except:
        raise Exception("Cannot JSON Serialise object: " + obj.__class__.__name__)
    return out

from json import JSONEncoder
_default.default = JSONEncoder.default
JSONEncoder.default = _default


class Server:
    @property
    def disks(self):
        if self.drives is None or self.images is None:
            self.update()
        all = self.drives + self.images
        return {'disk%d' % i: d for i, d in enumerate(all)}

    @property
    def host(self):
        return self.config['host']

    @property
    def port(self):
        return self.config['port']

    def update(self):
        if self.images is None or self.drives is None:
            changed = True
            h = 0
        else:
            changed = False
            h = hash(tuple(self.images))

        self.images = [DiskImage(f) for f in list_images(self.config['root'])]

        if hash(tuple(self.images)) != h:
            changed = True

        try:
            if changed is False:
                h = hash(tuple(self.drives))
            self.drives = [OpticalDrive(f) for f in list_optical_drives()]
            if changed is False and hash(tuple(self.drives)) != h:
                changed = True
        except:
            self.drives = []

        if 'Bonjour' in self.plugins.keys() and changed:
            self.plugins['Bonjour'].update()

    def __init__(self, config_file_path):
        self.config_file_path = config_file_path
        logging.info("Using config file %s" % self.config_file_path)

        self.config = {
            "host": get_local_ip(),
            "port": 49152,
            "root": "/mnt/images"
        }

        try:
            with open(self.config_file_path) as f:
                config = json.loads(f.read())
            self.config.update(config)
        except:
            self.save_config()

        if not os.path.exists(self.config['root']):
            # Create the images path if possible, otherwise raise an error
            # but continue gracefully.
            pass

        self.images = None
        self.drives = None

        cherrypy.config.update({
            'server.socket_host': self.host,
            'server.socket_port': self.port
        })

        self.load_plugins()

    def load_plugins(self):
        logging.info("Loading plugins")
        files = os.listdir(os.path.join(os.path.dirname(__file__), "Plugins"))
        self.plugins = {}
        for f in files:
            if f.startswith('__'): continue

            module_name = "pyods.Plugins." + f[:-3]
            try:
                module = importlib.import_module(module_name)
                plugin = getattr(module, str(module.__plugin__))
                logging.info("Loaded plugin named " + module_name)
                self.plugins[plugin.__name__] = plugin(self)
            except:
                logging.exception("Cannot load plugin named " + module_name)
                continue

    def run(self):
        logging.info("Starting webserver")
        cherrypy.engine.start()
        cherrypy.engine.block()

    def stop(self):
        self.save_config()
        self.plugins = []
        cherrypy.engine.stop()

    def save_config(self):
        if os.access(self.config_file_path, os.W_OK):
            config_data = json.dumps(self.config, indent=4, sort_keys=True)
            with open(self.config_file_path, 'w') as f:
                f.write(config_data)


def find_config():
    w = None
    for f in search_config:
        f = os.path.expanduser(f)
        d = os.path.dirname(f)
        if os.path.exists(d):
            if os.path.exists(f):
                return f
            if os.access(d, os.W_OK):
                w = f
        elif os.access(os.path.dirname(d), os.W_OK):
            os.makedirs(d)
            w = f
    if w is None:
        logging.error("Cannot find suitable config path to write, create on in %s" % ' or '.join(search_config))
        raise Exception("No config file")
    return w


def main():
    logging.basicConfig(
        level=logging.DEBUG,
        format='[%(asctime)s] %(levelname)s:thread-%(threadName)-9s:%(message)s'
    )
    logging.info("Starting pyods remote disk server")
    config = find_config()
    s = Server(config)
    s.run()

if __name__ == '__main__':
    main()


pyods/Plugins/Image.py
import cherrypy
import logging
import os

__plugin__ = "ImageAPI"


class ImageAPI:
    exposed = True
    path = '/images'
    config = {
        '/': {
            'request.dispatch': cherrypy.dispatch.MethodDispatcher(),
            'tools.json_out.on': True
        }
    }

    def __init__(self, server):
        self.server = server
        cherrypy.tree.mount(self, self.__class__.path, self.__class__.config)

    def POST(self, image):
        # Upload image to the image bucket
        size = 0
        success = False
        try:
            ext = os.path.splitext(image.filename)[1][1:].lower()
            if ext not in ['img', 'dmg', 'iso']:
                raise cherrypy.HTTPError(403, "Unsupported image file")
            path = os.path.join(self.server.config['root'], image.filename)
            with open(path, "wb") as f:
                while True:
                    data = image.file.read(8192)

                    if not data:
                        break

                    size += len(data)
                    f.write(data)
            self.server.update()
            success = True
        except:
            logging.exception("Failed to store upload")

        return {
            'size': size,
            'success': success
        }

    def GET(self):
        return self.server.disks

    def DELETE(self, image_name):
        # self.server.disks[image_name].erase()
        pass


pyods/Plugins/Bonjour.py
from zeroconf import ServiceInfo, Zeroconf
import socket
import logging

__plugin__ = "Bonjour"


class Bonjour:
    def __init__(self, server):
        self.server = server
        self.zeroconf = Zeroconf()
        self.info = None
        self.update()

    def update(self):
        self.remove()
        hostname = socket.gethostname()

        desc = {
            'sys': 'waMA=A4:BA:DB:E7:89:CD,adVF=0x4,adDT=0x3,adCC=1'
        }

        for (ident, disk) in self.server.disks.items():
            desc[ident] = 'adVN=%s,adVT=public.cd-media' % disk.label
            logging.info("Announcing disk \"%s\" as %s with name \"%s\"" % (disk.filename, ident, disk.label))

        self.info = ServiceInfo(
            "_odisk._tcp.local.",
            "%s._odisk._tcp.local." % hostname,
            addresses=[socket.inet_aton(self.server.host)],
            port=self.server.port,
            properties=desc,
        )

        self.add()

    def add(self):
        if self.info is None:
            return
        self.zeroconf.register_service(self.info)

    def remove(self):
        if self.info is None:
            return
        self.zeroconf.unregister_service(self.info)

    def __del__(self):
        self.remove()
        self.zeroconf.close()


pyods/Plugins/ODS.py
import cherrypy
import re
import os
from datetime import datetime
import logging
from pyods.OnlineDisk import OnlineDiskState

__plugin__ = "ODS"


class ODS:
    exposed = True
    path = '/'
    config = {
        '/': {
            'request.dispatch': cherrypy.dispatch.MethodDispatcher(),
            'tools.response_headers.on': True,
            'tools.response_headers.headers': [
                ('Content-Type', 'application/octet-stream'),
                ('Server', 'ODS/1.0')
            ]
        }
    }

    def __init__(self, server):
        self.server = server
        self._re_range = re.compile('bytes=(\d*?)-(\d*)')
        cherrypy.tree.mount(self, self.__class__.path, self.__class__.config)

    def GET(self, disk, fork=None, misc=None):
        """
        A get request is sent for reading a "chunk" of a disk using a supplied byte
        range.
        """
        if cherrypy.request.headers['user-agent'] != 'CCURLBS::readDataFork':
            logging.debug("User-Agent: " + str(cherrypy.request.headers['user-agent']))
            raise cherrypy.HTTPError(403)

        (disk, extension) = os.path.splitext(os.path.basename(disk))

        try:
            disk = self.server.disks[disk]
        except:
            raise cherrypy.HTTPError(404, "Disk not found")

        try:
            range = cherrypy.request.headers['range']
            start, end = self._re_range.findall(range)[0]
            start, end = int(start), int(end)
        except:
            logging.exception("range: " + str(range))
            raise cherrypy.HTTPError(500, "Range decode error: " + str(range))

        if disk.state == OnlineDiskState.READY:
            data = disk.read(start, end)
        else:
            raise cherrypy.HTTPError(404, 'Device not ready')

        cherrypy.response.headers['Content-Range'] = 'bytes %d-%d/%d' % (start, end, disk.size)
        return data

    def HEAD(self, disk, fork=None, misc=None):
        """
        Return the header indicating the disk size in bytes and the current date/time
        """
        if cherrypy.request.headers['user-agent'] != 'CCURLBS::statImage':
            logging.debug("User-Agent: " + str(cherrypy.request.headers['user-agent']))
            raise cherrypy.HTTPError(403)

        (disk, extension) = os.path.splitext(os.path.basename(disk))

        try:
            disk = self.server.disks[disk]
        except:
            raise cherrypy.HTTPError(404, "Disk not found")

        cherrypy.response.headers['Date'] = datetime.utcnow().isoformat()
        cherrypy.response.headers['Accept-Ranges'] = 'bytes'
        cherrypy.response.headers['Content-Length'] = disk.size
